src/replication/ng/registry/manager.go

```diff
@@ -15,20 +15,335 @@
+	return &DefaultManager{}
+}
+
+// Ensure *DefaultManager has implemented Manager interface.
+var _ Manager = (*DefaultManager)(nil)
+
+// Get gets a registry by id
+func (m *DefaultManager) Get(id int64) (*model.Registry, error) {
+	registry, err := dao.GetRegistry(id)
+	if err != nil {
+		return nil, err
+	}
+
+	if registry == nil {
+		return nil, fmt.Errorf("registry '%d' does not exist", id)
+	}
+
+	return fromDaoModel(registry)
+}
+
+// GetByName gets a registry by its name
+func (m *DefaultManager) GetByName(name string) (*model.Registry, error) {
+	registry, err := dao.GetRegistryByName(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if registry == nil {
+		return nil, nil
+	}
+
+	return fromDaoModel(registry)
+}
+
+// GetByURL gets a registry by its URL
+func (m *DefaultManager) GetByURL(url string) (*model.Registry, error) {
+	registry, err := dao.GetRegistryByURL(url)
+	if err != nil {
+		return nil, err
+	}
+
+	if registry == nil {
+		return nil, nil
+	}
+
+	return fromDaoModel(registry)
+}
+
+// List lists registries according to query provided.
+func (m *DefaultManager) List(query ...*model.RegistryQuery) (int64, []*model.Registry, error) {
+	var registryQueries []*dao.ListRegistryQuery
+	for _, q := range query {
+		// limit being -1 indicates no pagination specified, result in all registries matching name returned.
+		listQuery := &dao.ListRegistryQuery{
+			Query: q.Name,
+			Limit: -1,
+		}
+		if q.Pagination != nil {
+			listQuery.Offset = q.Pagination.Page * q.Pagination.Size
+			listQuery.Limit = q.Pagination.Size
+		}
+
+		registryQueries = append(registryQueries, listQuery)
+	}
+	total, registries, err := dao.ListRegistries(registryQueries...)
+	if err != nil {
+		return -1, nil, err
+	}
+
+	var results []*model.Registry
+	for _, r := range registries {
+		registry, err := fromDaoModel(r)
+		if err != nil {
+			return -1, nil, err
+		}
+		results = append(results, registry)
+	}
+
+	return total, results, nil
+}
+
+// Add adds a new registry
+func (m *DefaultManager) Add(registry *model.Registry) (int64, error) {
+	r, err := toDaoModel(registry)
+	if err != nil {
+		log.Errorf("Convert registry model to dao layer model error: %v", err)
+		return -1, err
+	}
+
+	id, err := dao.AddRegistry(r)
+	if err != nil {
+		log.Errorf("Add registry error: %v", err)
+		return -1, err
+	}
+
+	return id, nil
+}
+
+// Update updates a registry
+func (m *DefaultManager) Update(registry *model.Registry, props ...string) error {
+	// TODO(ChenDe): Only update the given props
+
+	r, err := toDaoModel(registry)
+	if err != nil {
+		log.Errorf("Convert registry model to dao layer model error: %v", err)
+		return err
+	}
+
+	return dao.UpdateRegistry(r)
+}
+
+// Remove deletes a registry
+func (m *DefaultManager) Remove(id int64) error {
+	policies, err := dao.GetRepPolicyByTarget(id)
+	if err != nil {
+		log.Errorf("Get policies related to registry %d error: %v", id, err)
+		return err
+	}
+
+	if len(policies) > 0 {
```

Comment:
I think the logic should be handle in the API level and use the global PolicyManager, it will be defined in the replication module

Reply:
That's better.